File: n_queens.cpp
class Solution {
public:
    bool finished(vector<string>& board, int n)
    {
        int count = 0;
        for(int i = 0; i < n; i++){
            for(int j = 0; j < n; j++){
                if(board[i][j] == 'Q'){
                    count++;
                }
            }
        }
    return count == n;
    }
    bool isSafe(vector<string> &board, int row, int col, int n)
    {
       for(int i = 0; i < n; i++){
            if(board[row][i] == 'Q' || board[i][col] == 'Q'){
                return false;
            }
        }
        
        for(int i = row, j = col; i >= 0 && j >= 0; i--, j--){
            if(board[i][j] =='Q'){
                return false;
            }
        }
        
        for(int i = row, j = col; i >=0 && j < n; i--, j++){
            if(board[i][j] =='Q'){
                return false;
            }            
        }
        return true;
    }
    void backtrack(vector<vector<string>> &res, vector<string> &board, int row, int n)
    {
        if(finished(board,n))
        {
            // cout<<"hello";
            // for(int i=0;i<n;i++)
            // {
            //     for(int j=0;j<n;j++)
            //         cout<<res[i][j]; 
            //         cout<<endl;
            // }
            res.push_back(board);
        }
        else
        {
            for(int j=0; j<n; j++)
            {
                if(isSafe(board,row,j,n))
                {
                    board[row][j] = 'Q';
                    backtrack(res,board,row+1,n);                      
                    board[row][j] = '.';
                }
            }
        }
    }
    vector<vector<string>> solveNQueens(int n) {
               
        vector<vector<string>> res;
        string line = "";
        for(int i = 0; i < n; i++){
            line+='.';
        }
        vector<string> board(n,line);
        backtrack(res,board,0,n);
        return res;
    }
};

File: NQueens.java
class Solution {

    List<List<String>> result = new ArrayList<List<String>>();

    public List<List<String>> solveNQueens(int n) {
        
        List<String> board = new ArrayList<String>();
        for(int i=0; i<n; i++) {
            StringBuilder row = new StringBuilder("");
            for(int j=0; j<n; j++)
                row.append(".");
            board.add(row.toString());
        }
        solve(0, board);

        return result;
    }

    public void solve(int row, List<String> board) {
        if(row == board.size()) {
            result.add(new ArrayList<>(board));
            return;
        }

        for(int j=0; j< board.size(); j++) {
            if(isSafe(row,j,board)) {
                StringBuilder newRow = new StringBuilder(board.get(row));
                newRow.setCharAt(j, 'Q');
                board.set(row, newRow.toString());
                solve(row+1, board);
                newRow.setCharAt(j, '.');
                board.set(row, newRow.toString());
            }
        }
    }

    public boolean isSafe(int row, int col, List<String> board) {
        // row-1, col
        // row-1, col-1
        // row-1, col+1
        for(int i=row-1; i>=0; i--) {
            if(board.get(i).charAt(col) == 'Q') return false;
        }
        for(int i=row-1, j=col-1; i>=0 && j>=0;) {
            if(board.get(i).charAt(j) == 'Q') return false;
            i--;
            j--;
        }
        for(int i=row-1, j=col+1; i>=0 && j<board.size();) {
            if(board.get(i).charAt(j) == 'Q') return false;
            i--;
            j++;
        }
        return true;
    }
}
